Backend/core/comms.py

```diff
@@ -0,0 +1,113 @@
+    return fields
+
+
+class TCP:
+    __tmp_data = b''
+
+    def listen_tcp(self, server_addr: str, port: int):
+        global solar_car_connection, frontend_data
+        while True:
+            # create a client socket
+            client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            try:
+                client.connect((server_addr, port))
+                solar_car_connection = True
+                print('connected')
+            except ConnectionRefusedError:
+                print(f'Connection to car server {server_addr} is refused')
+                continue
+
+            # Set the socket to non-blocking mode
+            client.setblocking(False)
+
+            # Create a list of sockets to monitor for incoming data
+            sockets = [client]
+
+            while True:
+                # Wait for data to be received or timeout after 5 seconds
+                readable, _, _ = select.select(sockets, [], [], 5)
+
+                if client in readable:
+                    data = client.recv(1000)
+                    if not data:
+                        # No data received, close the connection and break the loop
+                        client.close()
+                        break
+
+                    packets = self.parse_packets(data)
+                    for packet in packets:
+                        d = unpack_data(packet)
+                        frontend_data = d.copy()
+                        db.insert_data(d)
+                else:
+                    # Timeout occurred, close the connection and break the loop
+                    client.close()
+                    solar_car_connection = False
+                    break
+
+
+
+    def parse_packets(self, new_data: bytes):
+        """Parse and check the length of each packet"""
+        self.__tmp_data += new_data
+        packets = []
+
+        while True:
+            # Search for the next complete data packet
+            try:
+                start_index = self.__tmp_data.index(b'<bl>')
+                end_index = self.__tmp_data.index(b'</bl>')
+            except ValueError:
+                break
+
+            #print("start index:", start_index, "end index:", end_index)
+            #print(solar_car_connection)
+
+            # Extract a complete data packet
+            packets.append(self.__tmp_data[start_index + 4:end_index])
+            # Update the remaining data to exclude the processed packet
+            self.__tmp_data = self.__tmp_data[end_index + 5:]
+
+        # If the remaining data is longer than the expected packet length,
+        # there might be an incomplete packet, so log a warning.
+        if len(self.__tmp_data) >= byte_length:
+            print("ERROR: Incomplete or malformed packet ------------------------------------")
+            self.__tmp_data = b''
+
+        return packets
+
+
+def start_comms():
+    gen_format_str(config.DATAFORMAT_PATH)
+    tcp = TCP()
+    tcp.listen_tcp('127.0.0.1', 4003)
```

Comment:
```suggestion
    tcp.listen_tcp(TCP_ADDRESS, TCP_PORT)
```
Can we make 4003 a constant, as well as any other ports and addresses?

Reply:
ty